### PR TITLE
Eliminate insertFromComposition, deleteByComposition, and deleteCompositionText

### PR DESCRIPTION
**Pull request by BoCupp-Microsoft**
This PR eliminates three inputTypes from beforeinput/input events: insertFromComposition, deleteByComposition, and deleteCompositionText.

It also clarifies the order of the insertCompositionText beforeinput and input events with respect to composition events.  The proposed behavior is that beforeinput and input events for insertCompositionText occur only after compositionupdate events.

Additionally, a note stating that multiple compositionend events may occur for one compositionstart event was removed.

The result is that the composition sequence is:

* A compositionstart event
* Multiple occurrences of this sequence:
    * A compositionupdate event
    * A beforeinput event with inputType = insertCompositionText
    * An input event with inputType = insertCompositionText
* A compositionend event

***
<a href="https://pr-preview.s3.amazonaws.com/BoCupp-Microsoft/input-events/pull/122.html" title="Last updated on Sep 9, 2021, 6:59 AM UTC (288c62c)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/input-events/122/00c2127...BoCupp-Microsoft:288c62c.html" title="Last updated on Sep 9, 2021, 6:59 AM UTC (288c62c)">Diff</a>

***
<a href="https://pr-preview.s3.amazonaws.com/BoCupp-Microsoft/input-events/pull/122.html" title="Last updated on Mar 10, 2022, 2:49 PM UTC (288c62c)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/input-events/122/00c2127...BoCupp-Microsoft:288c62c.html" title="Last updated on Mar 10, 2022, 2:49 PM UTC (288c62c)">Diff</a>

### CODE REVIEW

**Comment by johanneswilm** [comment]
Hey, I don't think it would be a good idea to entirely remove them, because:

* They solve a fundamental problem of containerizing IME input
* It has been implemented in Webkit and this will then suddenly make Webkit not compliant with the spec.

I do understand that Chromium has some issues with implementing these features. The Chromium implementation is not able to solve the issues that arise from not implementing this part in any other way. However, the beforeinput/input event is still a plus for other uses.

Therefore, I would propose to make these events optional instead of removing them entirely.

**Comment by BoCupp-Microsoft** [comment]
Hi @johanneswilm, 

Sorry for the slow reply. 

The challenge as I see it isn’t in implementing the events, but in supporting a relocatable selection when composition starts. 

The goal of these events (per our previous conversations) is to enable an author to avoid having composition modify the portion of their DOM being used as the view for some editor-specific document model.  The author is meant to accomplish that goal by relocating selection and focus to some other part of the DOM where the author would like composition to occur, and then, as composition progresses, the author can incorporate the result back into their document model so they can render it themselves as part of their view – presumably back to the same place where selection was when composition first started. 

Let me know if you think I've stated the goal correctly and understood the way in which you expect the author to use these events. 

My mental model of how to enable this scenario looks like this: 

*Step 1*: During the compositionstart event, the author should move selection to a new editable element (call it the compositionElement) and remember where the selection was before the move (represent this location as a Range called originalCompositionRange). 

*Step 2*: For each compositionupdate event, the author should process the data field into their editor-specific document model and render it back into originalCompositionRange.  The author can repeatedly do this for every compositionupdate event, each time replacing the contents of the originalCompositionRange with new text.  Note that since composition is taking place in the compositionElement and not inside the originalCompositionRange, the author is free to produce whatever markup they want inside the originalCompositionRange, e.g. the author might surround every character with a span. 

*Step 3*: During the compositionend event, the author should move selection back to the end of originalCompositionRange and perform other necessary cleanup (like hiding the compositionElement). 

Note one challenge with step 1 is that it isn’t clear to me how much additional context the author should copy from the original editable element into the compositionElement.  On MacOS, the Japanese IME has a reconversion feature that expands the active composition range to the preceding phrase when the user presses left arrow.  In order for that feature to work, the phrase to be included in the expansion must be in the compositionElement.  Maybe the answer is to advise the author to copy a full paragraph of text? 

Another challenge with step 1 is that the selection should indicate the text to be reconverted.  This happens naturally when the text to reconvert is selected by the user.  Reconversion can also be triggered, however, when the selection is collapsed.  In this case the IME determines the range to reconvert.  Firefox appears to handle this case by setting the selection equal to the reconversion range prior to dispatching the compositionstart event. 

Some sample code showing the approach [can be found here](https://natural-seemly-limit.glitch.me/clone-content.html).  It almost works in Firefox on Windows - the first character typed in the composition is dropped as a result of moving the selection to the compositionElement, but it’s otherwise functional.  The first character issue I believe is unrelated to whether these events are kept or removed. 

It also almost works in Safari, but to facilitate reconversion and workaround selection not being equal to the reconversion range, I produced [another variant here](https://natural-seemly-limit.glitch.me/clone-with-range-from-insert.html) which derives the reconversion range from the first insertCompositionText beforeinput event.  This works for Safari on MacOS for all cases except when the user wants to expand the reconversion range using the left arrow.  For that scenario to work I further modified the code to this [final variant](https://natural-seemly-limit.glitch.me/clone-with-extra-events.html) which also handles insertFromComposition and deleteCompositionText by performing the same basic steps: replace the contents of originalCompositionRange with the data from the event. 

Besides Firefox on Windows and Safari on MacOS, other browsers have more significant problems with all variants. 

* Safari on iOS cancels the composition after moving the selection. 
* Chromium on Windows allows the composition to continue but drops the first character of the composition.  Reconversion based on a selection gets cancelled. 
* Chromium on MacOS allows the composition to continue but loses the first character of the composition.  Reconversion based on a selection starts a composition, but loses the context of what was being reconverted so that there are no suggestions from the candidate window. 
* Firefox on MacOS allows the composition to continue but loses the first character of the composition.  Reconversion based on a selection gets cancelled. 

In the [final variant](https://natural-seemly-limit.glitch.me/clone-with-extra-events.html) I mention above for Safari there are also side effects on the state of the IME where it is taken out of phrase mode when hitting the left arrow. 

Based on my observations from these experiments, my intuition is that there’s a long tail of issues to sort through both for the spec and implementations as we discover the sensitivities of various IMEs for different languages and different platforms.  When I consider that the EditContext accomplishes the same goal but doesn’t require a complex protocol to deal with focus changes during composition start and can also work for other scenarios like an editor that renders to a canvas – I'm reluctant to invest more in making a relocatable selection work. 

Could we instead declare that relocating selection during compositionstart is unsupported and remove these superflous events?

**Comment by johanneswilm** [comment]
Hey @BoCupp-Microsoft, thanks for experimenting. These events were created after discussions involving  tens of subject experts over an extended period of time. Moving the caret somewhere else is just one of multiple possible uses. Another is to leave the composition in its original place but to use the events at the end of the composition to do some editor-related cleanup of the dom. For example, an editor may want to add spans to mark who wrote a particular section of the document or to mark at which time it was written. Spans could also be used for a word-style tracked changes feature.
It is possible that EditContext will cover these cases. Have all three browser engines committed to fully implement EditContext within a reasonable timeframe? If not, I'd think it would be premature to remove the only currently working implementation of a spec that actually addresses the need to do this.
Why can we not do as we planned initially - mark these events as optional so that we can merge level 1 and 2 because Chromium has technical issues implementing this, but not remove them entirely so that webkit doesn't lose it's status as being compliant?
If you do think we should still go ahead and remove them entirely, can we have a meeting involving the subject experts that came up with the current solution first?

**Comment by BoCupp-Microsoft** [comment]
> Another is to leave the composition in its original place but to use the events at the end of the composition to do some editor-related cleanup of the dom. For example, an editor may want to add spans to mark who wrote a particular section of the document or to mark at which time it was written. Spans could also be used for a word-style tracked changes feature.

For this scenario what prevents authors from using compositionend?

> Why can we not do as we planned initially - mark these events as optional so that we can merge level 1 and 2 because Chromium has technical issues implementing this, but not remove them entirely so that webkit doesn't lose its status as being compliant?

That's fine.  I'll stop objecting.  Do you want to prepare that pull request and I'll abandon this one?

**Comment by johanneswilm** [comment]
> For this scenario what prevents authors from using compositionend?

In practice, compositionend, at times in combination with a timeout, is what is used in a lot fo editors today as it is one fo few events that is universally available. However, the event does not give enough information to make it clear where in the dom which changes were done. So with the event, one has to trigger something else - for example a dom diffing mechanism. That's what we attempted to get away from by giving the required information to the web dev so cleanup could be doing with diffing.
> That's fine. I'll stop objecting. Do you want to prepare that pull request and I'll abandon this one?

OK, I can do that.

**Comment by BoCupp-Microsoft** [comment]
@masayuki-nakano, @johanneswilm, @whsieh 

In this PR I was proposing we eliminate three input events that aren't implemented by Chromium browsers or Firefox.  I think the events are unnecessary, but the events are currently implemented in Safari.  [Johannes asked if we could mark the events optional](https://github.com/w3c/input-events/pull/122#issuecomment-944280491) instead of eliminating them.  [I said above that I wouldn't object](https://github.com/w3c/input-events/pull/122#issuecomment-944618755).

@whsieh and @masayuki-nakano could you let @johanneswilm and I know if you're happy with that outcome?  If so, I think we can cancel the meeting we were planning for this Friday to discuss how to move forward with this spec.  If you think we should discuss other options we can continue discussion here and in our ad-hoc meeting of the Web Editing WG this Friday.

Thanks for your thoughts!
Bo

**Comment by BoCupp-Microsoft** [comment]
Tagging related issue: https://github.com/w3c/input-events/issues/115

**Comment by masayuki-nakano** [comment]
I think that `deleteByComposition` and `deleteCompositionText` should not be fired (especially for `input`) from the draft (https://github.com/w3c/input-events/issues/86) because in my experience, changing dispatching event timing and/or count is risky change for existing web apps.

On the other hand, about `insertFromComposition`, I sometimes get feedback that web apps cannot cancel committing composition makes them require `MutationObserver` etc. But I think that it is not necessary if `insertCompositionText` can be cancelable only when committing the composition (https://github.com/w3c/input-events/issues/115) because I don't think that `inputType` is referred strictly from web apps except some special types like `insertFromPaste`.

So basically, I'd be happy if they'd be dropped from the draft because of the backward compatibility.

**Comment by johanneswilm** [comment]
> So basically, I'd be happy if they'd be dropped from the draft because of the backward compatibility.

The current proposal is that they are not dropped but made optional. Are you ok with that @masayuki-nakano?

**Comment by masayuki-nakano** [comment]
> > So basically, I'd be happy if they'd be dropped from the draft because of the backward compatibility.
> 
> The current proposal is that they are not dropped but made optional. Are you ok with that @masayuki-nakano?

Well, I don't like the unclear state, but it's fine to me (If Chromium would support them, I'd be not happy, but probably we'll follow it). On the other hand, I don't know whether such unclear state makes web developers confused or not. I guess they'd be confused... Probably, the spec should explain the reason in a note.

**Comment by johanneswilm** [comment]
> Well, I don't like the unclear state, but it's fine to me (If Chromium would support them, I'd be not happy, but probably we'll follow it). On the other hand, I don't know whether such unclear state makes web developers confused or not. I guess they'd be confused... Probably, the spec should explain the reason in a note.

Ok, will add a note.

**Comment by BoCupp-Microsoft** [comment]
> On the other hand, about `insertFromComposition`, I sometimes get feedback that web apps cannot cancel committing composition makes them require `MutationObserver` etc. But I think that it is not necessary if `insertCompositionText` can be cancelable only when committing the composition (#115) because I don't think that `inputType` is referred strictly from web apps except some special types like `insertFromPaste`.

I'd like to better understand this comment.  Why would cancellation only during commit be useful?  The DOM is modified after every `compositionupdate` / `insertCompositionText`.  Why put special requirements around the last one?

**Comment by masayuki-nakano** [comment]
> > On the other hand, about `insertFromComposition`, I sometimes get feedback that web apps cannot cancel committing composition makes them require `MutationObserver` etc. But I think that it is not necessary if `insertCompositionText` can be cancelable only when committing the composition (#115) because I don't think that `inputType` is referred strictly from web apps except some special types like `insertFromPaste`.
> 
> I'd like to better understand this comment. Why would cancellation only during commit be useful? The DOM is modified after every `compositionupdate` / `insertCompositionText`. Why put special requirements around the last one?

The merit of `beforeinput` is that it's possible to cancel user input before DOM tree update. Therefore they may want to cancel the final input from IME composition, e.g., perhaps for filtering characters. Currently other `insertFrom*` and `insertText` are canceleable, but not for any IME compositions. Therefore, web app developers need special handling only for IME. Therefore, this is unhappy case for both web developers and IME users.

Note that canceling "composing characters" may make serious a11y issue. E.g., cannot see characters which are required to convert, and composing character should not be filtered because there are some cases that composing characters are not acceptable but final string is acceptable, e.g., converting from CJK names to their emails. Therefore, `beforeinput` for composing state shouldn't be cancelable.

**Comment by BoCupp-Microsoft** [comment]
> The merit of beforeinput is that it's possible to cancel user input before DOM tree update. 

Agree.  What I don't agree with is that cancellation of just the final committed text is important.  If the author can't stop composition from modifying their DOM for the first N compositionupdate events, then being able to cancel N+1 doesn't seem to matter.  The author still needs to deal with cleaning up their DOM and can process the data of the last compositionupdate by either pulling the final composition string from the DOM or from the event's data property.  I can't see how being able to cancel just the last event helps.

**Comment by masayuki-nakano** [comment]
> What I don't agree with is that cancellation of just the final committed text is important. If the author can't stop composition from modifying their DOM for the first N compositionupdate events, then being able to cancel N+1 doesn't seem to matter.

Indeed, if web apps so sensitive for DOM mutations, you're right. But I **guess** some apps may not be so sensitive, but may want to inserting the string by themselves. I assume that canceling the final `beforeinput` event for a composition deletes composing text by web browser, not keeps it as-is.

**Comment by johanneswilm** [comment]
For reference, in this issue @chong-z explained why we added these events in the first place: https://github.com/w3c/input-events/issues/34

**Comment by johanneswilm** [comment]
@BoCupp-Microsoft I don't see us having decided on any action at the meeting. Is that also what you recall from the meeting? What do you propose we do at this stage given that the positions are the ones they are? As far as I can tell, currently it says that the events are required. AFAICT, there does not seem to be consensus for removing them, but there seems to be potential consensus for marking them optional. Is that also your understanding? We cancelled the meeting last week, but if there are still things to clarify, maybe we should need another special meeting?
I would think it would mainly make sense if there are new people or new views that potentially could take part in such a debate. Just having the same people mention the same points again seems less fruitful. In that case I'd say we should probably resolve to just keeping things as they are with two different levels of the spec until such time that something changes the circumstances.

**Comment by whsieh** [comment]
I think we (WebKit) would be okay with removing support for these three input event types, to align with Gecko and Blink.

**Comment by johanneswilm** [comment]
> think we (WebKit) would be

@whsieh OK, can we get confirmation on that that it is definitely OK so that we can go ahead and merge this.

**Comment by DavidMulder0** [comment]
As someone who develops a browser based structured document editor I am confused and disheartened seeing that `deleteByComposition` is going to get removed. Right now our editor is read-only in Android Chromium based browsers, *because* the `deleteByComposition` isn't supported. As far as we are aware there is literally no way to support the Chrome/Edge for Android recomposition flow at the moment, as there is no way to know what part of the DOM is being changed. We have mocked up certain prototypes that would diff the DOM before and after composition, but given the nature of our editor this has proven unfeasible to do in a generic way. The - as far as we were aware - upcoming `deleteByComposition` allowed us to promise our customers that Android support would be forthcoming eventually, but now it seems this change would move support to a 'maybe one day Chromium will support EditContext'.

We went through the same testing @BoCupp-Microsoft described where we tried to change the selection in response to the compositionstart event, but this (unsurprisingly) turned out to be mired with issues, some of which BoCupp described[1]. We settled on a solution where we allow the composition events to uncontrollably pollute our DOM, after which we rerender the entire section of the document when the composition finishes using the `InputEvent.data` to know what to insert. Without `deleteByComposition` this fails in the case of recomposition though, so basically from our point of view `deleteByComposition` is the crucial absolutely necessary glue to make Input Events have any meaning. This might sound harsh, but without it the entire spec seems to do more harm than good[2]. 

In conclusion: Maybe we're missing something (we well might be, mobile device support is low on our list of priorities, so most of the investigations have been done by a single person: me), but we really feel like these 3 events are crucial additions to the Input Events API. Even if EditContext will provide a solution that we would like to use one day, that shouldn't be a reason to handicap this spec indefinitely. 

[1] We also investigated an approach where we would consistently keep the selection 'outside' the actually editable document, but this turned out to result in too much unpredictable behavior with different IME solutions (although we're aware that Google Docs was able to get this to work well).
[2] Reason I am saying this is because the spec creates the *illusion* that a good editor can be created in the browser relatively easily. Next you spend years (literally) developing a product just to find out that you basically can't fully support IME input. Luckily IME support can be decently hacked at level 1 as long as no recomposition happens... but the fact that theoretically any IME might break your editor is... scary.

**Comment by masayuki-nakano** [comment]
> As far as we are aware there is literally no way to support the Chrome/Edge for Android recomposition flow at the moment, as there is no way to know what part of the DOM is being changed.

Cannot retrieve the range to be recomposed with `Selection` at `compositionstart` and/or `InputEvent.getTargetRanges` at `beforeinput` whose `inputType` is `insertCompositionText`? If not, I think that it's a bug of Blink.

**Comment by johanneswilm** [comment]
@BoCupp-Microsoft Do you have an answer to the above concern?

**Comment by DavidMulder0** [comment]
@masayuki-nakano Right now out of the subject matter a bit, but the main problem I am aware of is that IME's really do not like it when anybody 'messes' with their composition area. Even if Chrome would correctly provide `getTargetRanges` on `insertCompositionText` this wouldn't allow updating the underlying data model, as rerendering the DOM will cause many (if not all) IME's to end the composition session (as far as I understand that's the reason why `preventDefault` isn't supported on composition input events). Theoretically however one could maybe track and combine the `getTargetRanges` on the client to only do the update once the composition ends, basically replicating the `insertFromComposition` manually. 

Still though, this results in a situation where I think a lot of implementations are going to be incorrect, as `insertCompositionText` will need a very uniquely different implementation from all other input events. Plus it requires this 'odd' interplay between the composition events to determine which `insertCompositionText` was last, to correctly handle undo/redo actions (as a compositionupdate should probably not trigger its own undo action).

To be fair, it's all doable and when I wrote the last comment I somehow hadn't noticed that `getTargetRanges` actually should be set for composition related events as well, so thanks for pointing that out.

**Comment by johanneswilm** [comment]
@DavidMulder0 Your last paragraph makes it sound like you see less of a problem with this change now than previously. Is that correctly understood, or do you still view removing these three types permanently from the spec as problematic?

**Comment by BoCupp-Microsoft** [comment]
> BoCupp-Microsoft Do you have an answer to the above concern?

The first beforeinput insertCompositionText should have the range to be reconverted marked before the text is updated.  That makes reconversion the same as every other case from the author's perspective: the range of insertCompositionText is indicating what will be replaced.

I would also be OK with doing as @masayuki-nakano suggests and making the selection match that range before dispatching compositionstart.  We don't currently do that in Chromium, and I'm not sure its necessary given we'll report the same range in the following insertCompositionText event, but I'd be willing to make a change for interoperability.

**Comment by BoCupp-Microsoft** [comment]
@DavidMulder0 could you confirm if insertCompositionText would meet your needs as I've described it being used above?

**Comment by masayuki-nakano** [comment]
@DavidMulder0 
> @masayuki-nakano Right now out of the subject matter a bit, but the main problem I am aware of is that IME's really do not like it when anybody 'messes' with their composition area.

Yes. If native (i.e., OS's) IME API in **all** platforms works smarter, browsers can do better. However, for example, IME on Windodws handling with TSF may stop working until restarting browser if the composing string is unexpected.

> as rerendering the DOM will cause many (if not all) IME's to end the composition session (as far as I understand that's the reason why `preventDefault` isn't supported on composition input events).

What should happen when IME gets unexpected result depends on IME in any platforms. I.e., out of scope of browsers. Therefore, `preventDefault` during composition is hard to implement.

> Plus it requires this 'odd' interplay between the composition events to determine which `insertCompositionText` was last, to correctly handle undo/redo actions (as a compositionupdate should probably not trigger its own undo action).

For solving such non-smart API, I suggested `InputEvent.isComposing`, but there are some problems, see https://github.com/w3c/uievents/issues/202 for the detail.

Basically, while IME has composition string browser nor web app should not do anything which may surprise IME, e.g., sanitizing the **composing** string, moving to different position. When the composition gets committed, browser and/or web app should put the result into its undo stack. Then, IME features except undoing last commit feature should work.

Therefore, I think that the last `insertCompositionText` of a composition could be cancelable (https://github.com/w3c/input-events/issues/115).

@BoCupp-Microsoft
> I would also be OK with doing as @masayuki-nakano suggests and making the selection match that range before dispatching compositionstart.

IME for TSF on Windows do same thing. First, setting selection to replace, then, insert (replace) text. Therefore I'm surprised at why Blink does not do this on Windows. And IIRC, reconversion on macOS is also setting selection first. So, doing it is reasonable behavior to me...

**Comment by johanneswilm** [comment]
@DavidMulder0 @BoCupp-Microsoft @masayuki-nakano Do you feel that we have reached a conclusion here and that this PR can be merged as is, or do we need to modify it?

**Comment by BoCupp-Microsoft** [comment]
Apologies, I was thinking this was ready to merge because @masayuki-nakano and I were in agreement and I didn't see any objection from @DavidMulder0.  When I read back through this thread though I see [this comment](https://github.com/w3c/input-events/pull/122#issuecomment-951496962) from @masayuki-nakano that I didn't address:

>I assume that canceling the final `beforeinput` event for a composition deletes composing text by web browser, not keeps it as-is.

With this PR (specifically with the removal of `deleteCompositionText`) there is no operation that deletes the composed text as part of committing it.  If the author wants to remove the text from the DOM and then reinsert it themselves, I believe they are free to do so using the last reported range of the `beforeinput` event which precedes `compositionend`.

The code would look like this:

```javascript
const ce = document.querySelector("[contenteditable]")      

let lastRange = null
ce.addEventListener("beforeinput", e => {
  if (e.inputType != "insertCompositionText") {
    return
  }
  
  const ranges = e.getTargetRanges()
  if (ranges.length) {
    lastRange = ranges[0]
  }
})

ce.addEventListener("compositionend", e => {
  if (lastRange) {
    // StaticRange doesn't have a deleteContents, so convert to live Range
    const r = new Range()
    r.setStart(lastRange.startContainer, lastRange.startOffset)
    r.setEnd(lastRange.endContainer, lastRange.endOffset)
    r.deleteContents()
  }
})
```

There's a [demo here](https://natural-seemly-limit.glitch.me/delete-at-compositionend.html) that logs events so you can clearly see what's happening.  I tested it in FireFox with the in-box Japanese IME on Mac and Windows.  @masayuki-nakano, let me know if you agree to having the author handle the scenario you raised as I have shown above.  If so, then I don't think we need any additional behavior to remove the composed text on the final `beforeinput` preceding `compositionend` and we can merge this PR as is.

**Review by johanneswilm — APPROVED** [review]

**Comment by masayuki-nakano** [comment]
@BoCupp-Microsoft  Sorry for the delay to reply, I got lost because of too busy days due to a house moving. And thank you for pinging me, @johanneswilm.

> With this PR (specifically with the removal of `deleteCompositionText`) there is no operation that deletes the composed text as part of committing it. If the author wants to remove the text from the DOM and then reinsert it themselves, I believe they are free to do so using the last reported range of the `beforeinput` event which precedes `compositionend`.

From point of view of editor app developers, I agree with you, however, according to some bug reports to Firefox, some (non-editor apps') web developers want to use `beforeinput` lighter. Your idea requires web apps need to implement undo transaction by themselves. However, such developers want to do that instead, composition should be cancelable if the final value is not acceptable, but undo transactions should keep working.

> 
> The code would look like this:
> 
> ```js
> const ce = document.querySelector("[contenteditable]")      
> 
> let lastRange = null
> ce.addEventListener("beforeinput", e => {
>   if (e.inputType != "insertCompositionText") {
>     return
>   }
>   
>   const ranges = e.getTargetRanges()
>   if (ranges.length) {
>     lastRange = ranges[0]
>   }
> })
> 
> ce.addEventListener("compositionend", e => {
>   if (lastRange) {
>     // StaticRange doesn't have a deleteContents, so convert to live Range
>     const r = new Range()
>     r.setStart(lastRange.startContainer, lastRange.startOffset)
>     r.setEnd(lastRange.endContainer, lastRange.endOffset)
>     r.deleteContents()
>   }
> })
> ```
> 
> There's a [demo here](https://natural-seemly-limit.glitch.me/delete-at-compositionend.html) that logs events so you can clearly see what's happening. I tested it in FireFox with the in-box Japanese IME on Mac and Windows. @masayuki-nakano, let me know if you agree to having the author handle the scenario you raised as I have shown above. If so, then I don't think we need any additional behavior to remove the composed text on the final `beforeinput` preceding `compositionend` and we can merge this PR as is.

So as far as I've tested, the demo breaks undo/redo behavior at least on both Firefox and Chrome. Therefore, I think that final composition being cancelable is reasonable to me, and it's consistent with the other text insertion `inputType`s.

On the other hand, this PR is a reasonable size, and including only one purpose, "removing the unnecessary `inputType` values". It seems that discussing and updating the specification in another PR may be better.

**Comment by BoCupp-Microsoft** [comment]
Thank you for the feedback @masayuki-nakano.  I'll open a separate issue to discuss cancellation for the final composition per this comment and merge this PR:

> On the other hand, this PR is a reasonable size, and including only one purpose, "removing the unnecessary inputType values". It seems that discussing and updating the specification in another PR may be better.

**Comment by BoCupp-Microsoft** [comment]
Created [this issue](https://github.com/w3c/input-events/issues/134) to discuss whether the last `beforeinput` before `compositionend` should be cancellable per [this comment](https://github.com/w3c/input-events/pull/122#issuecomment-1064706372).
